plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -204,22 +248,58 @@ func (c *CloudWatch) Gather(acc telegraf.Accumulator) error {
 		return err
 	}
 
+	// get all of the possible queries so we can send groups of 100
+	// note: these are cached using metricCache's specs (when only namespace is defined)
+	queries, err := c.getDataQueries(metrics)
+	if err != nil {
+		return err
+	}
+
 	// limit concurrency or we can easily exhaust user connection limit
 	// see cloudwatch API request limits:
 	// http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/cloudwatch_limits.html
 	lmtr := limiter.NewRateLimiter(c.RateLimit, time.Second)
 	defer lmtr.Stop()
-	var wg sync.WaitGroup
-	wg.Add(len(metrics))
-	for _, m := range metrics {
+	wg := sync.WaitGroup{}
+
+	// create master list of results so aggregation works the best way.
+	results := []*cloudwatch.MetricDataResult{}
+
+	// loop through metrics and send groups of 100 at a time to gatherMetrics
+	// gatherMetrics(acc, metrics[min:max]...)
+	groups := len(queries) / 100
+	for i := 0; i < groups; i++ {
+		wg.Add(1)
 		<-lmtr.C
-		go func(inm *cloudwatch.Metric) {
+		go func(inm []*cloudwatch.MetricDataQuery) {
 			defer wg.Done()
-			acc.AddError(c.gatherMetric(acc, inm))
-		}(m)
+			result, err := c.gatherMetrics(c.getDataInputs(inm))
+			if err != nil {
+				acc.AddError(err)
+				// return
+			}
+			results = append(results, result...)
+
+		}(queries[i*100 : (i+1)*100])
 	}
+
+	// gather remainder (or initial) group
+	<-lmtr.C
+	wg.Add(1)
+	go func(inm []*cloudwatch.MetricDataQuery) {
+		defer wg.Done()
+		result, err := c.gatherMetrics(c.getDataInputs(inm))
+		if err != nil {
+			acc.AddError(err)
+			// return
+		}
+		results = append(results, result...)
+	}(queries[(groups * 100):])
```

Comment:
I think remainder case can be a problem if len(queries) % 100 == 0.

Reply:
ooh, you are right

Reply:
Huh, surprisingly no.
```go
fmt.Println(strings.Repeat("b", 100)[100:])
```
Is completely valid, must be due to the ranging.